src/execution/mod.rs

```diff
@@ -42,10 +75,19 @@ impl Pump {
     fn pump(&mut self) {
         Python::with_gil(|py| {
             let mut pull_from_pyiter = self.pull_from_pyiter.0.as_ref(py);
-            if let Some(epoch_item_pytuple) = pull_from_pyiter.next() {
-                let (epoch, item) = with_traceback!(py, epoch_item_pytuple?.extract());
-                self.push_to_timely.advance_to(epoch);
-                self.push_to_timely.send(item);
+            if let Some(epoch_item) = pull_from_pyiter.next() {
+                match epoch_item {
+                    Ok(item) => {
+                        if let Ok(send) = item.downcast::<PyCell<TdSend>>() {
+                            self.push_to_timely.send(send.borrow().item.clone());
+                        } else if let Ok(advance_to) = item.downcast::<PyCell<TdAdvanceTo>>() {
+                            self.push_to_timely.advance_to(advance_to.borrow().epoch);
+                        }
+                    }
+                    Err(err) => {
+                        std::panic::panic_any(err);
```

Comment:
Is there a way to use `with_traceback!` instead here? This will be a Python error and we'd probably still want to log it before throwing using the same machinery.